python/ray/serialization.py

```diff
@@ -194,6 +194,10 @@ def _deserialize_object(self, data, metadata, object_ref):
                 if data is None:
                     return b""
                 return data.to_pybytes()
+            elif metadata_fields[0] == ray_constants.OBJECT_METADATA_TYPE_ARROW:
+                import pyarrow as pa
+                reader = pa.BufferReader(data)
+                return pa.ipc.open_stream(reader).read_all()
```

Comment:
If I understand correctly, the intended use is only for writing objects from Java and reading from Python right? You wouldn't use this for Python->Python communication since pyarrow serialized objects are already zero copy.

Essentially this is "adding back" Arrow serialization support that we previous dropped to get its cross-language serialization advantages (cc @suquark @pcmoritz for thoughts here).

Reply:
Yes, this is only Java -> Python now, but we will also need Python -> Java. I'm OK with using pickle5 for python -> python, maybe we can have an option to swtich behavior.

Reply:
@ericl The serialization advantages you mentioned here is from msgpack instead of arrow, right? or I missed any context?

Reply:
@jovany-wang you're probably right. So this is supporting Arrow tables within msgpack? Not sure how the writer side is implemented (seems the code is not yet added to this PR).

Reply:
I think I bypassed msgpack(or the whole java serializer). Is it possible with msgpack?
The code I used to test is in raydp, so it is not included here. What I do is 1. get data in byte array from ArrowStreamWriter, 2. create a NativeRayObject using the data and metadata b"ARROW" 3. use putRaw to put it in to plasma.

Reply:
Then pass the objectId and owner address to python, and register there. With the change of python deserializer, I am able to ray.get the object ref and it is of type pyarrow.table.

Reply:
It's definitely possible with msgpack. For example, there's a msgpack extension to serialize ray objectrefs specially already, we could do the same for arrow.